{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import os\r\n",
    "\r\n",
    "# change the working directory to where the example files are located\r\n",
    "os.chdir(os.path.realpath('alignment_stat_examples'))\r\n",
    "print(os.listdir())"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "['483510-hisat2-hybrid-alignment-trimmed.out', '483513-hisat2-hybrid-alignment.out']\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "# import 483510-hisat2-hybrid-alignment-trimmed.out as a list with the first 117 lines trimmed out\r\n",
    "with open(os.listdir()[0], 'r') as f:\r\n",
    "    text = f.read().split(\"\\n\") # splitting each line and storing it as an element an a list\r\n",
    "    text = text[118:] # the first 117 lines are junk\r\n",
    "\r\n",
    "print(text[0:5])"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "['Aligning reads to the indexed genome using HISAT2 and generating sorted BAM files using Samtools...', 'amelRNA_1_1.trimmed.fq.gz', 'amelRNA_1_2.trimmed.fq.gz', '===============================================================================', 'Time loading forward index: 00:00:01']\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "# a function to extract sample names from the list that contains each new line in the text as an element\r\n",
    "def get_sample_names(list):\r\n",
    "    id = []\r\n",
    "    for i in range(0, len(list)): # looping through each line\r\n",
    "        if list[i] == \"Aligning reads to the indexed genome using HISAT2 and generating sorted BAM files using Samtools...\":\r\n",
    "            bam = list[i+1] # store the line after list[i]\r\n",
    "            # if the first line after list[i] contains the suffix \"trimmed.fq.gz\"\r\n",
    "            if bam[12:] == \".trimmed.fq.gz\" or bam[12:] == \"trimmed.fq.gz\":\r\n",
    "                id.append(bam.replace(\"_1.trimmed.fq.gz\", \"\")) # append the prefix (sample ID)\r\n",
    "            # if the first line after list[i] contains the suffix \"fq.gz\"\r\n",
    "            elif bam[12:] == \".fq.gz\" or bam[12:] == \"fq.gz\":\r\n",
    "                id.append(bam.replace(\"_1.fq.gz\", \"\")) # append the prefix (sample ID)\r\n",
    "    return id\r\n",
    "\r\n",
    "sample_ids = get_sample_names(text)\r\n",
    "print(sample_ids)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "['amelRNA_1', 'amelRNA_2', 'amelRNA_3', 'amelRNA_4', 'amelRNA_5', 'amelRNA_6', 'amelRNA_7', 'amelRNA_8', 'amelRNA_9', 'amelRNA_10', 'amelRNA_11', 'amelRNA_12', 'amelRNA_13', 'amelRNA_14', 'amelRNA_15', 'amelRNA_16', 'amelRNA_18', 'amelRNA_19']\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "# a function to extract alignment statistics\r\n",
    "def get_alignment_stats(list):\r\n",
    "    alignment_stats = []\r\n",
    "\r\n",
    "    # parts of lines (suffix) where the info wanted is found to be used to extract them\r\n",
    "    # since the info we want is found before the parts assigned below in the same line, we refer to them as suffix\r\n",
    "    reads_all = \"reads; of these:\" # total number of paired reads\r\n",
    "    conc_one = \"aligned concordantly exactly 1 time\" # number of paired reads aligned concordantly exactly 1 time\r\n",
    "    conc_multi = \"aligned concordantly >1 times\" # number of paired reads aligned concordantly exactly more than 1 time\r\n",
    "    reads_disconc = \"pairs aligned concordantly 0 times; of these:\" # number of paired reads aligned concordantly exactly 0 time s\r\n",
    "    disconc_one = \"aligned discordantly 1 time\" # number of reads aligned discordantly 1 time\r\n",
    "    pair_mates = \"pairs aligned 0 times concordantly or discordantly; of these:\" # number of paired reads aligned 0 times concordantly or discordantly\r\n",
    "    mates_zero = \"mates make up the pairs; of these:\" # number of reads that make up the pairs\r\n",
    "    mates_one = \"aligned exactly 1 time\" # number of individual reads (mates) aligned exactly 1 time\r\n",
    "    mates_multi = \"aligned >1 times\" # number of individual reads (mates) aligned exactly more than 1 time\r\n",
    "    overall = \"overall alignment rate\" # overall alignment rate of reads to the genome\r\n",
    "\r\n",
    "    for element in list: # looping through the lines\r\n",
    "        # finding the index suffixes start\r\n",
    "        reads_all_ind = element.find(reads_all)\r\n",
    "        conc_one_ind = element.find(conc_one)\r\n",
    "        conc_multi_ind = element.find(conc_multi)\r\n",
    "        reads_disconc_ind = element.find(reads_disconc)\r\n",
    "        disconc_one_ind = element.find(disconc_one)\r\n",
    "        pair_mates_ind = element.find(pair_mates)\r\n",
    "        mates_zero_ind = element.find(mates_zero)\r\n",
    "        mates_one_ind = element.find(mates_one)\r\n",
    "        mates_multi_ind = element.find(mates_multi)\r\n",
    "        overall_ind = element.find(overall)\r\n",
    "        \r\n",
    "        # if a line contains one of the suffixes specificed above, append the prefixes (information of interest) to a new list\r\n",
    "        # removing any leading and trailing whitespaces\r\n",
    "        if element[reads_all_ind:] == reads_all:\r\n",
    "            alignment_stats.append(element[:reads_all_ind].strip())\r\n",
    "        elif element[conc_one_ind:] == conc_one:\r\n",
    "            alignment_stats.append(element[:conc_one_ind].strip())\r\n",
    "        elif element[conc_multi_ind:] == conc_multi:\r\n",
    "            alignment_stats.append(element[:conc_multi_ind].strip())\r\n",
    "        elif element[reads_disconc_ind:] == reads_disconc:\r\n",
    "            alignment_stats.append(element[:reads_disconc_ind].strip())\r\n",
    "        elif element[disconc_one_ind:] == disconc_one:\r\n",
    "            alignment_stats.append(element[:disconc_one_ind].strip())\r\n",
    "        elif element[pair_mates_ind:] == pair_mates:\r\n",
    "            alignment_stats.append(element[:pair_mates_ind].strip())\r\n",
    "        elif element[mates_zero_ind:] == mates_zero:\r\n",
    "            alignment_stats.append(element[:mates_zero_ind].strip())\r\n",
    "        elif element[mates_one_ind:] == mates_one:\r\n",
    "            alignment_stats.append(element[:mates_one_ind].strip())\r\n",
    "        elif element[mates_multi_ind:] == mates_multi:\r\n",
    "            alignment_stats.append(element[:mates_multi_ind].strip())\r\n",
    "        elif element[overall_ind:] == overall:\r\n",
    "            alignment_stats.append(element[:overall_ind].strip())\r\n",
    "    return alignment_stats\r\n",
    "\r\n",
    "alignment_stats = get_alignment_stats(text)\r\n",
    "print(alignment_stats)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "['29620775', '26049478 (87.94%)', '525302 (1.77%)', '3045995', '225090 (7.39%)', '2820905', '5641810', '1780566 (31.56%)', '56423 (1.00%)', '93.58%', '30911037', '27239267 (88.12%)', '680166 (2.20%)', '2991604', '274388 (9.17%)', '2717216', '5434432', '1792804 (32.99%)', '72653 (1.34%)', '94.23%', '31395825', '27971795 (89.09%)', '572134 (1.82%)', '2851896', '272897 (9.57%)', '2578999', '5157998', '1602742 (31.07%)', '85290 (1.65%)', '94.47%', '32112536', '28430984 (88.54%)', '535309 (1.67%)', '3146243', '228136 (7.25%)', '2918107', '5836214', '1931165 (33.09%)', '76512 (1.31%)', '94.04%', '32081947', '28367725 (88.42%)', '637570 (1.99%)', '3076652', '279069 (9.07%)', '2797583', '5595166', '1874079 (33.49%)', '66125 (1.18%)', '94.30%', '31950608', '28587787 (89.47%)', '460095 (1.44%)', '2902726', '276220 (9.52%)', '2626506', '5253012', '1735708 (33.04%)', '55250 (1.05%)', '94.58%', '30718531', '27126080 (88.31%)', '396309 (1.29%)', '3196142', '395726 (12.38%)', '2800416', '5600832', '1788322 (31.93%)', '67819 (1.21%)', '93.90%', '30008714', '26617522 (88.70%)', '527332 (1.76%)', '2863860', '265331 (9.26%)', '2598529', '5197058', '1751189 (33.70%)', '43633 (0.84%)', '94.33%', '30743620', '27284363 (88.75%)', '505743 (1.65%)', '2953514', '323062 (10.94%)', '2630452', '5260904', '1675953 (31.86%)', '47650 (0.91%)', '94.25%', '31864855', '27449748 (86.14%)', '528560 (1.66%)', '3886547', '248791 (6.40%)', '3637756', '7275512', '2134107 (29.33%)', '67312 (0.93%)', '92.04%', '30573422', '25319222 (82.81%)', '448284 (1.47%)', '4805916', '275312 (5.73%)', '4530604', '9061208', '1671759 (18.45%)', '58426 (0.64%)', '88.01%', '32571115', '28939493 (88.85%)', '381349 (1.17%)', '3250273', '405557 (12.48%)', '2844716', '5689432', '1671312 (29.38%)', '58861 (1.03%)', '93.92%', '31652897', '28288823 (89.37%)', '467531 (1.48%)', '2896543', '266376 (9.20%)', '2630167', '5260334', '1716597 (32.63%)', '69590 (1.32%)', '94.51%', '29628405', '25693821 (86.72%)', '621517 (2.10%)', '3313067', '387492 (11.70%)', '2925575', '5851150', '1835352 (31.37%)', '77597 (1.33%)', '93.35%', '31708754', '27764888 (87.56%)', '531288 (1.68%)', '3412578', '396345 (11.61%)', '3016233', '6032466', '1853281 (30.72%)', '67921 (1.13%)', '93.52%', '29853017', '25344208 (84.90%)', '531541 (1.78%)', '3977268', '350158 (8.80%)', '3627110', '7254220', '1603781 (22.11%)', '60550 (0.83%)', '90.64%', '31176380', '27650410 (88.69%)', '616289 (1.98%)', '2909681', '313336 (10.77%)', '2596345', '5192690', '1698229 (32.70%)', '71732 (1.38%)', '94.51%', '30699600', '26890929 (87.59%)', '560815 (1.83%)', '3247856', '389719 (12.00%)', '2858137', '5716274', '1690236 (29.57%)', '73503 (1.29%)', '93.56%']\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "source": [
    "# a function to build nested dictionaries containins alignment rates/number of reads for samples\r\n",
    "def alignment_stats_dict(stats, sample_list):\r\n",
    "    # keys for dictionaries that will serve as column names\r\n",
    "    h0_1 = \"Pairs aligned concordantly\"\r\n",
    "    h0_2 = \"Pairs aligned disconcordantly\"\r\n",
    "    h0_3 = \"Pairs aligned 0 times concordantly or discordantly\"\r\n",
    "    h1_1 = \"Total number of pairs\"\r\n",
    "    h1_2 = \"Total number of mates\"\r\n",
    "    h1_3 = \"Uniquely aligned pairs (%)\"\r\n",
    "    h1_4 = \"Uniquely aligned mates (%)\"\r\n",
    "    h1_5 = \"Multimapped pairs (%)\"\r\n",
    "    h1_6 = \"Multimapped mates (%)\"\r\n",
    "    h2 = \"\"\r\n",
    "    h2_1 = \"Overall alignment rate\"\r\n",
    "\r\n",
    "    # intializing dictionaries for each key (column)\r\n",
    "    ls1_1, ls1_2, ls1_3, ls1_4, ls1_5 = {}, {}, {}, {}, {}\r\n",
    "    ls1_6, ls1_7, ls1_8, ls1_9, ls1_10 = {}, {}, {}, {}, {}\r\n",
    "\r\n",
    "    for id in sample_list: # looping through sample ID list and intilizing dictionaries with sample names\r\n",
    "        ls1_1[id], ls1_2[id], ls1_3[id], ls1_4[id], ls1_5[id] = 0, 0, 0, 0, 0\r\n",
    "        ls1_6[id], ls1_7[id], ls1_8[id], ls1_9[id], ls1_10[id] = 0, 0, 0, 0, 0\r\n",
    "\r\n",
    "    # initializing dictionaries with the key specified above\r\n",
    "    h1_1_dic = dict({h1_1:0, h1_3:0, h1_5:0})\r\n",
    "    h1_2_dic = dict({h1_1:0, h1_3:0})\r\n",
    "    h1_3_dic = dict({h1_1:0, h1_2:0, h1_4:0, h1_6:0})\r\n",
    "    h2_dic = dict({h2_1:0})\r\n",
    "    # merging everything into one dictionary\r\n",
    "    h0_dic = dict({h0_1:h1_1_dic, h0_2:h1_2_dic, h0_3:h1_3_dic, h2:h2_dic})\r\n",
    "\r\n",
    "    # looping through the length or stats (should be 180 element long) with a step of 10 \r\n",
    "    # this will allow us to loop through every single quantity for each sample, filling the intilized dictionaries with sample names\r\n",
    "    for i in range(0, len(stats), 10):\r\n",
    "        # sample_list should be 18 element long as we have 18 samples\r\n",
    "        key = sample_list[int(i/10)]\r\n",
    "        # for each sample, append the quantity\r\n",
    "        ls1_1[key] = stats[i] \r\n",
    "        ls1_2[key] = stats[i+1]\r\n",
    "        ls1_3[key] = stats[i+2]\r\n",
    "        ls1_4[key] = stats[i+3]\r\n",
    "        ls1_5[key] = stats[i+4]\r\n",
    "        ls1_6[key] = stats[i+5]\r\n",
    "        ls1_7[key] = stats[i+6]\r\n",
    "        ls1_8[key] = stats[i+7]\r\n",
    "        ls1_9[key] = stats[i+8]\r\n",
    "        ls1_10[key] = stats[i+9]\r\n",
    "\r\n",
    "    # piecing everything together under the appropriate headers\r\n",
    "    h0_dic[h0_1][h1_1] = ls1_1\r\n",
    "    h0_dic[h0_1][h1_3] = ls1_2\r\n",
    "    h0_dic[h0_1][h1_5] = ls1_3\r\n",
    "    h0_dic[h0_2][h1_1] = ls1_4\r\n",
    "    h0_dic[h0_2][h1_3] = ls1_5\r\n",
    "    h0_dic[h0_3][h1_1] = ls1_6\r\n",
    "    h0_dic[h0_3][h1_2] = ls1_7\r\n",
    "    h0_dic[h0_3][h1_4] = ls1_8\r\n",
    "    h0_dic[h0_3][h1_6] = ls1_9\r\n",
    "    h0_dic[h2][h2_1] = ls1_10\r\n",
    "    \r\n",
    "    return h0_dic\r\n",
    "\r\n",
    "stats_dict = alignment_stats_dict(alignment_stats, sample_ids)\r\n",
    "print(stats_dict)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "{'Pairs aligned concordantly': {'Total number of pairs': {'amelRNA_1': '29620775', 'amelRNA_2': '30911037', 'amelRNA_3': '31395825', 'amelRNA_4': '32112536', 'amelRNA_5': '32081947', 'amelRNA_6': '31950608', 'amelRNA_7': '30718531', 'amelRNA_8': '30008714', 'amelRNA_9': '30743620', 'amelRNA_10': '31864855', 'amelRNA_11': '30573422', 'amelRNA_12': '32571115', 'amelRNA_13': '31652897', 'amelRNA_14': '29628405', 'amelRNA_15': '31708754', 'amelRNA_16': '29853017', 'amelRNA_18': '31176380', 'amelRNA_19': '30699600'}, 'Uniquely aligned pairs (%)': {'amelRNA_1': '26049478 (87.94%)', 'amelRNA_2': '27239267 (88.12%)', 'amelRNA_3': '27971795 (89.09%)', 'amelRNA_4': '28430984 (88.54%)', 'amelRNA_5': '28367725 (88.42%)', 'amelRNA_6': '28587787 (89.47%)', 'amelRNA_7': '27126080 (88.31%)', 'amelRNA_8': '26617522 (88.70%)', 'amelRNA_9': '27284363 (88.75%)', 'amelRNA_10': '27449748 (86.14%)', 'amelRNA_11': '25319222 (82.81%)', 'amelRNA_12': '28939493 (88.85%)', 'amelRNA_13': '28288823 (89.37%)', 'amelRNA_14': '25693821 (86.72%)', 'amelRNA_15': '27764888 (87.56%)', 'amelRNA_16': '25344208 (84.90%)', 'amelRNA_18': '27650410 (88.69%)', 'amelRNA_19': '26890929 (87.59%)'}, 'Multimapped pairs (%)': {'amelRNA_1': '525302 (1.77%)', 'amelRNA_2': '680166 (2.20%)', 'amelRNA_3': '572134 (1.82%)', 'amelRNA_4': '535309 (1.67%)', 'amelRNA_5': '637570 (1.99%)', 'amelRNA_6': '460095 (1.44%)', 'amelRNA_7': '396309 (1.29%)', 'amelRNA_8': '527332 (1.76%)', 'amelRNA_9': '505743 (1.65%)', 'amelRNA_10': '528560 (1.66%)', 'amelRNA_11': '448284 (1.47%)', 'amelRNA_12': '381349 (1.17%)', 'amelRNA_13': '467531 (1.48%)', 'amelRNA_14': '621517 (2.10%)', 'amelRNA_15': '531288 (1.68%)', 'amelRNA_16': '531541 (1.78%)', 'amelRNA_18': '616289 (1.98%)', 'amelRNA_19': '560815 (1.83%)'}}, 'Pairs aligned disconcordantly': {'Total number of pairs': {'amelRNA_1': '3045995', 'amelRNA_2': '2991604', 'amelRNA_3': '2851896', 'amelRNA_4': '3146243', 'amelRNA_5': '3076652', 'amelRNA_6': '2902726', 'amelRNA_7': '3196142', 'amelRNA_8': '2863860', 'amelRNA_9': '2953514', 'amelRNA_10': '3886547', 'amelRNA_11': '4805916', 'amelRNA_12': '3250273', 'amelRNA_13': '2896543', 'amelRNA_14': '3313067', 'amelRNA_15': '3412578', 'amelRNA_16': '3977268', 'amelRNA_18': '2909681', 'amelRNA_19': '3247856'}, 'Uniquely aligned pairs (%)': {'amelRNA_1': '225090 (7.39%)', 'amelRNA_2': '274388 (9.17%)', 'amelRNA_3': '272897 (9.57%)', 'amelRNA_4': '228136 (7.25%)', 'amelRNA_5': '279069 (9.07%)', 'amelRNA_6': '276220 (9.52%)', 'amelRNA_7': '395726 (12.38%)', 'amelRNA_8': '265331 (9.26%)', 'amelRNA_9': '323062 (10.94%)', 'amelRNA_10': '248791 (6.40%)', 'amelRNA_11': '275312 (5.73%)', 'amelRNA_12': '405557 (12.48%)', 'amelRNA_13': '266376 (9.20%)', 'amelRNA_14': '387492 (11.70%)', 'amelRNA_15': '396345 (11.61%)', 'amelRNA_16': '350158 (8.80%)', 'amelRNA_18': '313336 (10.77%)', 'amelRNA_19': '389719 (12.00%)'}}, 'Pairs aligned 0 times concordantly or discordantly': {'Total number of pairs': {'amelRNA_1': '2820905', 'amelRNA_2': '2717216', 'amelRNA_3': '2578999', 'amelRNA_4': '2918107', 'amelRNA_5': '2797583', 'amelRNA_6': '2626506', 'amelRNA_7': '2800416', 'amelRNA_8': '2598529', 'amelRNA_9': '2630452', 'amelRNA_10': '3637756', 'amelRNA_11': '4530604', 'amelRNA_12': '2844716', 'amelRNA_13': '2630167', 'amelRNA_14': '2925575', 'amelRNA_15': '3016233', 'amelRNA_16': '3627110', 'amelRNA_18': '2596345', 'amelRNA_19': '2858137'}, 'Total number of mates': {'amelRNA_1': '5641810', 'amelRNA_2': '5434432', 'amelRNA_3': '5157998', 'amelRNA_4': '5836214', 'amelRNA_5': '5595166', 'amelRNA_6': '5253012', 'amelRNA_7': '5600832', 'amelRNA_8': '5197058', 'amelRNA_9': '5260904', 'amelRNA_10': '7275512', 'amelRNA_11': '9061208', 'amelRNA_12': '5689432', 'amelRNA_13': '5260334', 'amelRNA_14': '5851150', 'amelRNA_15': '6032466', 'amelRNA_16': '7254220', 'amelRNA_18': '5192690', 'amelRNA_19': '5716274'}, 'Uniquely aligned mates (%)': {'amelRNA_1': '1780566 (31.56%)', 'amelRNA_2': '1792804 (32.99%)', 'amelRNA_3': '1602742 (31.07%)', 'amelRNA_4': '1931165 (33.09%)', 'amelRNA_5': '1874079 (33.49%)', 'amelRNA_6': '1735708 (33.04%)', 'amelRNA_7': '1788322 (31.93%)', 'amelRNA_8': '1751189 (33.70%)', 'amelRNA_9': '1675953 (31.86%)', 'amelRNA_10': '2134107 (29.33%)', 'amelRNA_11': '1671759 (18.45%)', 'amelRNA_12': '1671312 (29.38%)', 'amelRNA_13': '1716597 (32.63%)', 'amelRNA_14': '1835352 (31.37%)', 'amelRNA_15': '1853281 (30.72%)', 'amelRNA_16': '1603781 (22.11%)', 'amelRNA_18': '1698229 (32.70%)', 'amelRNA_19': '1690236 (29.57%)'}, 'Multimapped mates (%)': {'amelRNA_1': '56423 (1.00%)', 'amelRNA_2': '72653 (1.34%)', 'amelRNA_3': '85290 (1.65%)', 'amelRNA_4': '76512 (1.31%)', 'amelRNA_5': '66125 (1.18%)', 'amelRNA_6': '55250 (1.05%)', 'amelRNA_7': '67819 (1.21%)', 'amelRNA_8': '43633 (0.84%)', 'amelRNA_9': '47650 (0.91%)', 'amelRNA_10': '67312 (0.93%)', 'amelRNA_11': '58426 (0.64%)', 'amelRNA_12': '58861 (1.03%)', 'amelRNA_13': '69590 (1.32%)', 'amelRNA_14': '77597 (1.33%)', 'amelRNA_15': '67921 (1.13%)', 'amelRNA_16': '60550 (0.83%)', 'amelRNA_18': '71732 (1.38%)', 'amelRNA_19': '73503 (1.29%)'}}, '': {'Overall alignment rate': {'amelRNA_1': '93.58%', 'amelRNA_2': '94.23%', 'amelRNA_3': '94.47%', 'amelRNA_4': '94.04%', 'amelRNA_5': '94.30%', 'amelRNA_6': '94.58%', 'amelRNA_7': '93.90%', 'amelRNA_8': '94.33%', 'amelRNA_9': '94.25%', 'amelRNA_10': '92.04%', 'amelRNA_11': '88.01%', 'amelRNA_12': '93.92%', 'amelRNA_13': '94.51%', 'amelRNA_14': '93.35%', 'amelRNA_15': '93.52%', 'amelRNA_16': '90.64%', 'amelRNA_18': '94.51%', 'amelRNA_19': '93.56%'}}}\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "source": [
    "import pandas as pd\r\n",
    "\r\n",
    "def build_stats_df(stats_dict):\r\n",
    "    # column names\r\n",
    "    h0_1 = \"Pairs aligned concordantly\"\r\n",
    "    h0_2 = \"Pairs aligned disconcordantly\"\r\n",
    "    h0_3 = \"Pairs aligned 0 times concordantly or discordantly\"\r\n",
    "    h2 = \"\"\r\n",
    "\r\n",
    "    # building dataframes for each dictionary\r\n",
    "    df1 = pd.DataFrame.from_dict(stats_dict[h0_1], orient=\"index\")\r\n",
    "    df2 = pd.DataFrame.from_dict(stats_dict[h0_2], orient=\"index\")\r\n",
    "    df3 = pd.DataFrame.from_dict(stats_dict[h0_3], orient=\"index\")\r\n",
    "    df4 = pd.DataFrame.from_dict(stats_dict[h2], orient=\"index\")\r\n",
    "\r\n",
    "    pieces = {h0_1:df1, h0_2:df2, h0_3:df3, h2: df4} # merging dataframes into one dictionary \r\n",
    "    \r\n",
    "    result_dataframe = pd.concat(pieces) # piecing together everything as a dataframe\r\n",
    "    return result_dataframe\r\n",
    "\r\n",
    "stats_df = build_stats_df(stats_dict)\r\n",
    "stats_df"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>amelRNA_1</th>\n",
       "      <th>amelRNA_2</th>\n",
       "      <th>amelRNA_3</th>\n",
       "      <th>amelRNA_4</th>\n",
       "      <th>amelRNA_5</th>\n",
       "      <th>amelRNA_6</th>\n",
       "      <th>amelRNA_7</th>\n",
       "      <th>amelRNA_8</th>\n",
       "      <th>amelRNA_9</th>\n",
       "      <th>amelRNA_10</th>\n",
       "      <th>amelRNA_11</th>\n",
       "      <th>amelRNA_12</th>\n",
       "      <th>amelRNA_13</th>\n",
       "      <th>amelRNA_14</th>\n",
       "      <th>amelRNA_15</th>\n",
       "      <th>amelRNA_16</th>\n",
       "      <th>amelRNA_18</th>\n",
       "      <th>amelRNA_19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Pairs aligned concordantly</th>\n",
       "      <th>Total number of pairs</th>\n",
       "      <td>29620775</td>\n",
       "      <td>30911037</td>\n",
       "      <td>31395825</td>\n",
       "      <td>32112536</td>\n",
       "      <td>32081947</td>\n",
       "      <td>31950608</td>\n",
       "      <td>30718531</td>\n",
       "      <td>30008714</td>\n",
       "      <td>30743620</td>\n",
       "      <td>31864855</td>\n",
       "      <td>30573422</td>\n",
       "      <td>32571115</td>\n",
       "      <td>31652897</td>\n",
       "      <td>29628405</td>\n",
       "      <td>31708754</td>\n",
       "      <td>29853017</td>\n",
       "      <td>31176380</td>\n",
       "      <td>30699600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniquely aligned pairs (%)</th>\n",
       "      <td>26049478 (87.94%)</td>\n",
       "      <td>27239267 (88.12%)</td>\n",
       "      <td>27971795 (89.09%)</td>\n",
       "      <td>28430984 (88.54%)</td>\n",
       "      <td>28367725 (88.42%)</td>\n",
       "      <td>28587787 (89.47%)</td>\n",
       "      <td>27126080 (88.31%)</td>\n",
       "      <td>26617522 (88.70%)</td>\n",
       "      <td>27284363 (88.75%)</td>\n",
       "      <td>27449748 (86.14%)</td>\n",
       "      <td>25319222 (82.81%)</td>\n",
       "      <td>28939493 (88.85%)</td>\n",
       "      <td>28288823 (89.37%)</td>\n",
       "      <td>25693821 (86.72%)</td>\n",
       "      <td>27764888 (87.56%)</td>\n",
       "      <td>25344208 (84.90%)</td>\n",
       "      <td>27650410 (88.69%)</td>\n",
       "      <td>26890929 (87.59%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Multimapped pairs (%)</th>\n",
       "      <td>525302 (1.77%)</td>\n",
       "      <td>680166 (2.20%)</td>\n",
       "      <td>572134 (1.82%)</td>\n",
       "      <td>535309 (1.67%)</td>\n",
       "      <td>637570 (1.99%)</td>\n",
       "      <td>460095 (1.44%)</td>\n",
       "      <td>396309 (1.29%)</td>\n",
       "      <td>527332 (1.76%)</td>\n",
       "      <td>505743 (1.65%)</td>\n",
       "      <td>528560 (1.66%)</td>\n",
       "      <td>448284 (1.47%)</td>\n",
       "      <td>381349 (1.17%)</td>\n",
       "      <td>467531 (1.48%)</td>\n",
       "      <td>621517 (2.10%)</td>\n",
       "      <td>531288 (1.68%)</td>\n",
       "      <td>531541 (1.78%)</td>\n",
       "      <td>616289 (1.98%)</td>\n",
       "      <td>560815 (1.83%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Pairs aligned disconcordantly</th>\n",
       "      <th>Total number of pairs</th>\n",
       "      <td>3045995</td>\n",
       "      <td>2991604</td>\n",
       "      <td>2851896</td>\n",
       "      <td>3146243</td>\n",
       "      <td>3076652</td>\n",
       "      <td>2902726</td>\n",
       "      <td>3196142</td>\n",
       "      <td>2863860</td>\n",
       "      <td>2953514</td>\n",
       "      <td>3886547</td>\n",
       "      <td>4805916</td>\n",
       "      <td>3250273</td>\n",
       "      <td>2896543</td>\n",
       "      <td>3313067</td>\n",
       "      <td>3412578</td>\n",
       "      <td>3977268</td>\n",
       "      <td>2909681</td>\n",
       "      <td>3247856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniquely aligned pairs (%)</th>\n",
       "      <td>225090 (7.39%)</td>\n",
       "      <td>274388 (9.17%)</td>\n",
       "      <td>272897 (9.57%)</td>\n",
       "      <td>228136 (7.25%)</td>\n",
       "      <td>279069 (9.07%)</td>\n",
       "      <td>276220 (9.52%)</td>\n",
       "      <td>395726 (12.38%)</td>\n",
       "      <td>265331 (9.26%)</td>\n",
       "      <td>323062 (10.94%)</td>\n",
       "      <td>248791 (6.40%)</td>\n",
       "      <td>275312 (5.73%)</td>\n",
       "      <td>405557 (12.48%)</td>\n",
       "      <td>266376 (9.20%)</td>\n",
       "      <td>387492 (11.70%)</td>\n",
       "      <td>396345 (11.61%)</td>\n",
       "      <td>350158 (8.80%)</td>\n",
       "      <td>313336 (10.77%)</td>\n",
       "      <td>389719 (12.00%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Pairs aligned 0 times concordantly or discordantly</th>\n",
       "      <th>Total number of pairs</th>\n",
       "      <td>2820905</td>\n",
       "      <td>2717216</td>\n",
       "      <td>2578999</td>\n",
       "      <td>2918107</td>\n",
       "      <td>2797583</td>\n",
       "      <td>2626506</td>\n",
       "      <td>2800416</td>\n",
       "      <td>2598529</td>\n",
       "      <td>2630452</td>\n",
       "      <td>3637756</td>\n",
       "      <td>4530604</td>\n",
       "      <td>2844716</td>\n",
       "      <td>2630167</td>\n",
       "      <td>2925575</td>\n",
       "      <td>3016233</td>\n",
       "      <td>3627110</td>\n",
       "      <td>2596345</td>\n",
       "      <td>2858137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total number of mates</th>\n",
       "      <td>5641810</td>\n",
       "      <td>5434432</td>\n",
       "      <td>5157998</td>\n",
       "      <td>5836214</td>\n",
       "      <td>5595166</td>\n",
       "      <td>5253012</td>\n",
       "      <td>5600832</td>\n",
       "      <td>5197058</td>\n",
       "      <td>5260904</td>\n",
       "      <td>7275512</td>\n",
       "      <td>9061208</td>\n",
       "      <td>5689432</td>\n",
       "      <td>5260334</td>\n",
       "      <td>5851150</td>\n",
       "      <td>6032466</td>\n",
       "      <td>7254220</td>\n",
       "      <td>5192690</td>\n",
       "      <td>5716274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniquely aligned mates (%)</th>\n",
       "      <td>1780566 (31.56%)</td>\n",
       "      <td>1792804 (32.99%)</td>\n",
       "      <td>1602742 (31.07%)</td>\n",
       "      <td>1931165 (33.09%)</td>\n",
       "      <td>1874079 (33.49%)</td>\n",
       "      <td>1735708 (33.04%)</td>\n",
       "      <td>1788322 (31.93%)</td>\n",
       "      <td>1751189 (33.70%)</td>\n",
       "      <td>1675953 (31.86%)</td>\n",
       "      <td>2134107 (29.33%)</td>\n",
       "      <td>1671759 (18.45%)</td>\n",
       "      <td>1671312 (29.38%)</td>\n",
       "      <td>1716597 (32.63%)</td>\n",
       "      <td>1835352 (31.37%)</td>\n",
       "      <td>1853281 (30.72%)</td>\n",
       "      <td>1603781 (22.11%)</td>\n",
       "      <td>1698229 (32.70%)</td>\n",
       "      <td>1690236 (29.57%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Multimapped mates (%)</th>\n",
       "      <td>56423 (1.00%)</td>\n",
       "      <td>72653 (1.34%)</td>\n",
       "      <td>85290 (1.65%)</td>\n",
       "      <td>76512 (1.31%)</td>\n",
       "      <td>66125 (1.18%)</td>\n",
       "      <td>55250 (1.05%)</td>\n",
       "      <td>67819 (1.21%)</td>\n",
       "      <td>43633 (0.84%)</td>\n",
       "      <td>47650 (0.91%)</td>\n",
       "      <td>67312 (0.93%)</td>\n",
       "      <td>58426 (0.64%)</td>\n",
       "      <td>58861 (1.03%)</td>\n",
       "      <td>69590 (1.32%)</td>\n",
       "      <td>77597 (1.33%)</td>\n",
       "      <td>67921 (1.13%)</td>\n",
       "      <td>60550 (0.83%)</td>\n",
       "      <td>71732 (1.38%)</td>\n",
       "      <td>73503 (1.29%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Overall alignment rate</th>\n",
       "      <td>93.58%</td>\n",
       "      <td>94.23%</td>\n",
       "      <td>94.47%</td>\n",
       "      <td>94.04%</td>\n",
       "      <td>94.30%</td>\n",
       "      <td>94.58%</td>\n",
       "      <td>93.90%</td>\n",
       "      <td>94.33%</td>\n",
       "      <td>94.25%</td>\n",
       "      <td>92.04%</td>\n",
       "      <td>88.01%</td>\n",
       "      <td>93.92%</td>\n",
       "      <td>94.51%</td>\n",
       "      <td>93.35%</td>\n",
       "      <td>93.52%</td>\n",
       "      <td>90.64%</td>\n",
       "      <td>94.51%</td>\n",
       "      <td>93.56%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                       amelRNA_1  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                29620775   \n",
       "                                                   Uniquely aligned pairs (%)  26049478 (87.94%)   \n",
       "                                                   Multimapped pairs (%)          525302 (1.77%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3045995   \n",
       "                                                   Uniquely aligned pairs (%)     225090 (7.39%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2820905   \n",
       "                                                   Total number of mates                 5641810   \n",
       "                                                   Uniquely aligned mates (%)   1780566 (31.56%)   \n",
       "                                                   Multimapped mates (%)           56423 (1.00%)   \n",
       "                                                   Overall alignment rate                 93.58%   \n",
       "\n",
       "                                                                                       amelRNA_2  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                30911037   \n",
       "                                                   Uniquely aligned pairs (%)  27239267 (88.12%)   \n",
       "                                                   Multimapped pairs (%)          680166 (2.20%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 2991604   \n",
       "                                                   Uniquely aligned pairs (%)     274388 (9.17%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2717216   \n",
       "                                                   Total number of mates                 5434432   \n",
       "                                                   Uniquely aligned mates (%)   1792804 (32.99%)   \n",
       "                                                   Multimapped mates (%)           72653 (1.34%)   \n",
       "                                                   Overall alignment rate                 94.23%   \n",
       "\n",
       "                                                                                       amelRNA_3  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                31395825   \n",
       "                                                   Uniquely aligned pairs (%)  27971795 (89.09%)   \n",
       "                                                   Multimapped pairs (%)          572134 (1.82%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 2851896   \n",
       "                                                   Uniquely aligned pairs (%)     272897 (9.57%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2578999   \n",
       "                                                   Total number of mates                 5157998   \n",
       "                                                   Uniquely aligned mates (%)   1602742 (31.07%)   \n",
       "                                                   Multimapped mates (%)           85290 (1.65%)   \n",
       "                                                   Overall alignment rate                 94.47%   \n",
       "\n",
       "                                                                                       amelRNA_4  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                32112536   \n",
       "                                                   Uniquely aligned pairs (%)  28430984 (88.54%)   \n",
       "                                                   Multimapped pairs (%)          535309 (1.67%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3146243   \n",
       "                                                   Uniquely aligned pairs (%)     228136 (7.25%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2918107   \n",
       "                                                   Total number of mates                 5836214   \n",
       "                                                   Uniquely aligned mates (%)   1931165 (33.09%)   \n",
       "                                                   Multimapped mates (%)           76512 (1.31%)   \n",
       "                                                   Overall alignment rate                 94.04%   \n",
       "\n",
       "                                                                                       amelRNA_5  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                32081947   \n",
       "                                                   Uniquely aligned pairs (%)  28367725 (88.42%)   \n",
       "                                                   Multimapped pairs (%)          637570 (1.99%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3076652   \n",
       "                                                   Uniquely aligned pairs (%)     279069 (9.07%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2797583   \n",
       "                                                   Total number of mates                 5595166   \n",
       "                                                   Uniquely aligned mates (%)   1874079 (33.49%)   \n",
       "                                                   Multimapped mates (%)           66125 (1.18%)   \n",
       "                                                   Overall alignment rate                 94.30%   \n",
       "\n",
       "                                                                                       amelRNA_6  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                31950608   \n",
       "                                                   Uniquely aligned pairs (%)  28587787 (89.47%)   \n",
       "                                                   Multimapped pairs (%)          460095 (1.44%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 2902726   \n",
       "                                                   Uniquely aligned pairs (%)     276220 (9.52%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2626506   \n",
       "                                                   Total number of mates                 5253012   \n",
       "                                                   Uniquely aligned mates (%)   1735708 (33.04%)   \n",
       "                                                   Multimapped mates (%)           55250 (1.05%)   \n",
       "                                                   Overall alignment rate                 94.58%   \n",
       "\n",
       "                                                                                       amelRNA_7  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                30718531   \n",
       "                                                   Uniquely aligned pairs (%)  27126080 (88.31%)   \n",
       "                                                   Multimapped pairs (%)          396309 (1.29%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3196142   \n",
       "                                                   Uniquely aligned pairs (%)    395726 (12.38%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2800416   \n",
       "                                                   Total number of mates                 5600832   \n",
       "                                                   Uniquely aligned mates (%)   1788322 (31.93%)   \n",
       "                                                   Multimapped mates (%)           67819 (1.21%)   \n",
       "                                                   Overall alignment rate                 93.90%   \n",
       "\n",
       "                                                                                       amelRNA_8  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                30008714   \n",
       "                                                   Uniquely aligned pairs (%)  26617522 (88.70%)   \n",
       "                                                   Multimapped pairs (%)          527332 (1.76%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 2863860   \n",
       "                                                   Uniquely aligned pairs (%)     265331 (9.26%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2598529   \n",
       "                                                   Total number of mates                 5197058   \n",
       "                                                   Uniquely aligned mates (%)   1751189 (33.70%)   \n",
       "                                                   Multimapped mates (%)           43633 (0.84%)   \n",
       "                                                   Overall alignment rate                 94.33%   \n",
       "\n",
       "                                                                                       amelRNA_9  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                30743620   \n",
       "                                                   Uniquely aligned pairs (%)  27284363 (88.75%)   \n",
       "                                                   Multimapped pairs (%)          505743 (1.65%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 2953514   \n",
       "                                                   Uniquely aligned pairs (%)    323062 (10.94%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2630452   \n",
       "                                                   Total number of mates                 5260904   \n",
       "                                                   Uniquely aligned mates (%)   1675953 (31.86%)   \n",
       "                                                   Multimapped mates (%)           47650 (0.91%)   \n",
       "                                                   Overall alignment rate                 94.25%   \n",
       "\n",
       "                                                                                      amelRNA_10  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                31864855   \n",
       "                                                   Uniquely aligned pairs (%)  27449748 (86.14%)   \n",
       "                                                   Multimapped pairs (%)          528560 (1.66%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3886547   \n",
       "                                                   Uniquely aligned pairs (%)     248791 (6.40%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 3637756   \n",
       "                                                   Total number of mates                 7275512   \n",
       "                                                   Uniquely aligned mates (%)   2134107 (29.33%)   \n",
       "                                                   Multimapped mates (%)           67312 (0.93%)   \n",
       "                                                   Overall alignment rate                 92.04%   \n",
       "\n",
       "                                                                                      amelRNA_11  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                30573422   \n",
       "                                                   Uniquely aligned pairs (%)  25319222 (82.81%)   \n",
       "                                                   Multimapped pairs (%)          448284 (1.47%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 4805916   \n",
       "                                                   Uniquely aligned pairs (%)     275312 (5.73%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 4530604   \n",
       "                                                   Total number of mates                 9061208   \n",
       "                                                   Uniquely aligned mates (%)   1671759 (18.45%)   \n",
       "                                                   Multimapped mates (%)           58426 (0.64%)   \n",
       "                                                   Overall alignment rate                 88.01%   \n",
       "\n",
       "                                                                                      amelRNA_12  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                32571115   \n",
       "                                                   Uniquely aligned pairs (%)  28939493 (88.85%)   \n",
       "                                                   Multimapped pairs (%)          381349 (1.17%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3250273   \n",
       "                                                   Uniquely aligned pairs (%)    405557 (12.48%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2844716   \n",
       "                                                   Total number of mates                 5689432   \n",
       "                                                   Uniquely aligned mates (%)   1671312 (29.38%)   \n",
       "                                                   Multimapped mates (%)           58861 (1.03%)   \n",
       "                                                   Overall alignment rate                 93.92%   \n",
       "\n",
       "                                                                                      amelRNA_13  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                31652897   \n",
       "                                                   Uniquely aligned pairs (%)  28288823 (89.37%)   \n",
       "                                                   Multimapped pairs (%)          467531 (1.48%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 2896543   \n",
       "                                                   Uniquely aligned pairs (%)     266376 (9.20%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2630167   \n",
       "                                                   Total number of mates                 5260334   \n",
       "                                                   Uniquely aligned mates (%)   1716597 (32.63%)   \n",
       "                                                   Multimapped mates (%)           69590 (1.32%)   \n",
       "                                                   Overall alignment rate                 94.51%   \n",
       "\n",
       "                                                                                      amelRNA_14  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                29628405   \n",
       "                                                   Uniquely aligned pairs (%)  25693821 (86.72%)   \n",
       "                                                   Multimapped pairs (%)          621517 (2.10%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3313067   \n",
       "                                                   Uniquely aligned pairs (%)    387492 (11.70%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2925575   \n",
       "                                                   Total number of mates                 5851150   \n",
       "                                                   Uniquely aligned mates (%)   1835352 (31.37%)   \n",
       "                                                   Multimapped mates (%)           77597 (1.33%)   \n",
       "                                                   Overall alignment rate                 93.35%   \n",
       "\n",
       "                                                                                      amelRNA_15  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                31708754   \n",
       "                                                   Uniquely aligned pairs (%)  27764888 (87.56%)   \n",
       "                                                   Multimapped pairs (%)          531288 (1.68%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3412578   \n",
       "                                                   Uniquely aligned pairs (%)    396345 (11.61%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 3016233   \n",
       "                                                   Total number of mates                 6032466   \n",
       "                                                   Uniquely aligned mates (%)   1853281 (30.72%)   \n",
       "                                                   Multimapped mates (%)           67921 (1.13%)   \n",
       "                                                   Overall alignment rate                 93.52%   \n",
       "\n",
       "                                                                                      amelRNA_16  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                29853017   \n",
       "                                                   Uniquely aligned pairs (%)  25344208 (84.90%)   \n",
       "                                                   Multimapped pairs (%)          531541 (1.78%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3977268   \n",
       "                                                   Uniquely aligned pairs (%)     350158 (8.80%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 3627110   \n",
       "                                                   Total number of mates                 7254220   \n",
       "                                                   Uniquely aligned mates (%)   1603781 (22.11%)   \n",
       "                                                   Multimapped mates (%)           60550 (0.83%)   \n",
       "                                                   Overall alignment rate                 90.64%   \n",
       "\n",
       "                                                                                      amelRNA_18  \\\n",
       "Pairs aligned concordantly                         Total number of pairs                31176380   \n",
       "                                                   Uniquely aligned pairs (%)  27650410 (88.69%)   \n",
       "                                                   Multimapped pairs (%)          616289 (1.98%)   \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 2909681   \n",
       "                                                   Uniquely aligned pairs (%)    313336 (10.77%)   \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2596345   \n",
       "                                                   Total number of mates                 5192690   \n",
       "                                                   Uniquely aligned mates (%)   1698229 (32.70%)   \n",
       "                                                   Multimapped mates (%)           71732 (1.38%)   \n",
       "                                                   Overall alignment rate                 94.51%   \n",
       "\n",
       "                                                                                      amelRNA_19  \n",
       "Pairs aligned concordantly                         Total number of pairs                30699600  \n",
       "                                                   Uniquely aligned pairs (%)  26890929 (87.59%)  \n",
       "                                                   Multimapped pairs (%)          560815 (1.83%)  \n",
       "Pairs aligned disconcordantly                      Total number of pairs                 3247856  \n",
       "                                                   Uniquely aligned pairs (%)    389719 (12.00%)  \n",
       "Pairs aligned 0 times concordantly or discordantly Total number of pairs                 2858137  \n",
       "                                                   Total number of mates                 5716274  \n",
       "                                                   Uniquely aligned mates (%)   1690236 (29.57%)  \n",
       "                                                   Multimapped mates (%)           73503 (1.29%)  \n",
       "                                                   Overall alignment rate                 93.56%  "
      ]
     },
     "metadata": {},
     "execution_count": 11
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# exporting the dataframe\r\n",
    "stats_df.to_csv(os.listdir()[0].replace(\".out\", \"\") + \".csv\")"
   ],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "language_info": {
   "name": "python",
   "version": "3.9.2",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.9.2 64-bit"
  },
  "interpreter": {
   "hash": "6d35480b5c1d294382b0e0921d54eccdfa53dc59be22530d71db640e0bbad820"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}